dev/AppNotifications/AppNotificationUtility.cpp

```diff
@@ -237,14 +240,101 @@ std::wstring Microsoft::Windows::AppNotifications::Helpers::GetDisplayNameBasedO
     return displayName;
 }
 
-// Placeholder
-HRESULT RetrieveAssetsFromProcess(_Out_ Microsoft::Windows::AppNotifications::Helpers::AppNotificationAssets& /*assets*/)
+#define GIL_EXACTSIZEONLY       0x0100
+// Looks at the passed in icon and tries loading a better sized icon from the resource file.
+// If there is a larger icon available, use that icon instead. This is because the HICON passed
+// in to the legacy tray item is either a SM_CXSMICON x SM_CYSMICON (16x16) icon, or if NIIF_LARGE_ICON
+// is specified in the dwInfoFlags then it should be a SM_CXICON x SM_CYICON (32x32) icon. For toast messages,
+// an image of 150x150 is expected. To allow a larger icon to be used we will check the icon sizes which are available.
+// When we take DPI into account, we always want to pick the next larger icon size and scale down. For example,
+// if we are running 200%, we are trying to load an image of 300x300 px. The next largest snap icon size is 512px,
+// but icon file format requires this to be available. So the largest possible size in an icon is 256px, which means
+// we should just load that image file size and ignore DPI.
+inline HICON LoadBetterIconFromResource(_In_ HICON hIcon, int cx, int cy, UINT uFlags /*= 0*/)
 {
-    // THROW_HR_IF_MSG(E_UNEXPECTED, VerifyIconFileExtension(iconFilePath));
+    HICON hIconNew = nullptr;
 
-    return E_NOTIMPL;
+    ICONINFOEX iconInfoEx = {};
+    iconInfoEx.cbSize = sizeof(iconInfoEx);
+    if (GetIconInfoEx(hIcon, &iconInfoEx))
+    {
+        if (S_OK != SHDefExtractIcon(iconInfoEx.szModName, -iconInfoEx.wResID, uFlags, &hIconNew, nullptr, MAKELONG(cx, cy)))
+        {
+            DestroyIcon(hIconNew);
+            hIconNew = nullptr;
+        }
+
+        THROW_IF_WIN32_BOOL_FALSE(DeleteObject(iconInfoEx.hbmColor));
+        THROW_IF_WIN32_BOOL_FALSE(DeleteObject(iconInfoEx.hbmMask));
+    }
+    return hIconNew;
 }
 
+inline HICON RetrieveIcon()
+{
+    std::wstring processPath{};
+    THROW_IF_FAILED(wil::GetModuleFileNameExW(GetCurrentProcess(), nullptr, processPath));
```

Comment:
Yes this API is only for unpackaged applications. Yes if the unpackaged apps don't ship an icon in the exe then we fallback to other mechanisms.

Note: This PR is draft and I wanted to get few things sorted out before it can be published however, thanks for the feedback so far @riverar . :)

Reply:
Noted will improvise it after tomorrow's discussion.